### DISCUSSION TITLE
Ability for buttons or menu items to fire events from where it is created

### DISCUSSION THREAD

**Discussion by JustRed23** [discord-jda/JDA#1795] [Questions and Help] [answered]
So I've had my issues with buttons for a while. I always need to send the specific button id with everything i need to be picked up by the `onButtonClick(@NonNull ButtonClickEvent event)` method and executed from there. My question is. Wouldn't it be nicer if you could fire the events from the button itself?

Example: 
`Button.primary("your-button-id", "your-label", ((ButtonClickEvent) event) -> System.out.println("execute code here"));`

I don't know if this would be a good idea but i have implemented that into my bot and it works fine.
Anyways thank you for checking out my issue.

**Comment by JustRed23** [comment]
I will link my version [here](https://github.com/JustRed23/Grandfather) so you can check out what i mean.

**Comment by MinnDevelopment** [comment, marked as answer]
This is not possible since events are dispatched from the gateway, that cannot be accessed by the button factory methods at all. If you want something like that you can make your own utility class which abstracts the listener filtering process for you, or use something like [jda-ktx](https://github.com/MinnDevelopment/jda-ktx/) which offers the `jda.onButton("id") { ... }` handler.

You can always add a listener at runtime with [JDA#addEventListener](https://ci.dv8tion.net/job/JDA/javadoc/net/dv8tion/jda/api/JDA.html#addEventListener(java.lang.Object...)):

```java
var button = Button.secondary("id", "Label");
event.getJDA().addEventListener(buttonListener("id", (e) -> ...)); // some custom listener code
event.getChannel().sendMessage("Click the button").setActionRow(button).queue(); // send the button
```

**Reply by JustRed23** [reply]
Okay! Thank you very much for answering!
